spec/Candid.md

```diff
@@ -608,6 +770,23 @@ The premise means that the rule does not apply when the constituent type is itse
 
 Q: The negated nature of this premise isn't really compatible with parametric polymorphism. Is that a problem? We could always introduce a supertype of all non-nullable types and rephrase it with that.
 
+Finally, in order to maintain *transitivity* of subtyping, an unusual rule allows, in fact, *any* option type to be regarded as a subtype of any other.
+```
+not (<datatype> <: <datatype'>)
+---------------------------------
+opt <datatype> <: opt <datatype'>
+```
+*Note:* This rule is necessary in the presence of the unusual record and variant rules shown below. Without it, certain upgrades may generally be valid one step at a time, but not taken together, which could cause problems for clients catching up with multiple upgrades.
+For example, given a record type `record {666 : opt nat}` it is valid to remove the field `666` by the rule below and evolve the type to `record {}`.
+A later step might legally re-add a field of the same name but with a different type, producing, e.g.,`record {666 : opt text}`.
+A client having missed the intermediate step will have to upgrade directly from the original to the newest version of the type.
+If two option type do not match up, its value will be treated as `null`.
+
+In practice, users are strongly discouraged to ever remove a record field or a variant tag and later re-add it with a different meaning.
```

Comment:
Wouldn’t this work:
```suggestion
In practice, users are strongly discouraged to ever remove an optional record field or a variant tag and later re-add it with a different meaning. Instead of removing an optional record field, it should be replaced with `opt empty`, to prevent re-usd of that field.
```
Maybe we can introduce an alias for `opt empty` to make this less obscure?

This is also how I envisioned the restricted “sensible evolution” relation to look like: It would not allow removing optional fields, only “disabling” them in this way.

This might require a form of polarities in the rules for “can evolve to”. So why do I think this might work (when it didn’t work for normal subtyping)? Because in the higher order case, where two services evolve independently, the composition needs to work with regard to the unrestricted subtyping relation, which is larger.

Reply:
Hm, I'm not convinced that wouldn't run afoul all the same problems in the higher-order or third-party cases. But happy to explore, later. :)